### ISSUE TITLE
An event should be able to have more than one location

### ISSUE THREAD

**Issue by KyFaSt** [railsbridge/bridge_troll#320]
Our installfest and workshop are hosted at two different locations, it would be excellent if we could denote that on the event page.

It would be doubly helpful if the generated ical events (which are awesome) could include the respective addresses too.

If someone can give me some guidance, I'm 100% invested in working on this feature myself!

**Comment by tjgrathwell** [comment]
As far as I can tell, this involves at least these pieces of work:

## Allowing organizers to set different locations for sessions

Currently session creation looks like this:
![screen shot 2015-04-26 at 12 17 06 am](https://cloud.githubusercontent.com/assets/1041691/7336024/a34a5ee0-ebab-11e4-805f-153d7d70a7eb.png)

One approach is to keep the existing per-event `location`, and allow each `event_session` to have its own `location` (or maybe `location_override`, or `custom_location`) which could be defined like this:
![screen shot 2015-04-26 at 12 15 19 am](https://cloud.githubusercontent.com/assets/1041691/7336025/a8861fac-ebab-11e4-8b03-186f62704ff5.png)

Alternatively, we could push `location` off `events` entirely and down to `event_sessions`, which is maybe more conceptually pure, but could complicate some of the other logic in the app (and might be optimizing for the uncommon case).

## Fixing emails / ical

Currently the emails sent to attendees look like this:
![screen shot 2015-04-26 at 12 25 36 am](https://cloud.githubusercontent.com/assets/1041691/7336032/9f7a92c0-ebac-11e4-9f51-6cc6792d9f65.png)

Since that only mentions a single location, we might need to break it out per-session like this:
![screen shot 2015-04-26 at 12 29 08 am](https://cloud.githubusercontent.com/assets/1041691/7336034/b0989476-ebac-11e4-9e1f-4fc986b68b86.png)

It's an open question whether to preserve the simpler format when all sessions are at the same place. It might look silly to keep repeating the same location details over and over.

For ical, since the link is per-session already it should be pretty simple to include the correct location details once the organizer is able to define them.

## Fixing the event page

This is the one I'm least confident about. The event page design is still heavily modeled after Meetup, which emphasizes a single event location.

Currently the event page looks like this:
![screen shot 2015-04-26 at 12 17 35 am](https://cloud.githubusercontent.com/assets/1041691/7336042/054847f0-ebad-11e4-89fb-ef068a55d647.png)

There's only space for the one location listed, so if each session has its own unique location it might need to look like this:
![screen shot 2015-04-26 at 12 23 04 am](https://cloud.githubusercontent.com/assets/1041691/7336043/1e3ed6e8-ebad-11e4-9026-5fba46bf585c.png)

...but then what should that map lurking on the side be displaying? The "main location"? Should there be two little maps? A ground-up redesign? What if there's three sessions (usually Teacher Training / Installfest / Workshop)? I just don't know.

SO, what are your thoughts about
* How should organizers be entering the location in this multi-location universe (on every session, or with one "main" location that is overridden for certain sessions?)
* What should the event page look like when sessions have different locations? How do attendees know where to go at what time?

**Comment by KyFaSt** [comment]
Super thorough @tjgrathwell!

I think I follow what you're saying.
* I think it makes sense for each `event_session` to default to to the `event`'s `location` unless some kind of `location_override` is selected in the view. I think an implementation could look like this:
  - The `location_override` could be a JavaScript toggle that displays a location selectbox containing potential `custom_location`s
  - A `custom_location` column would be added to the `event_sessions` table

* While I want this to be as easy as possible for users, I also don't want to complicate the view logic for an edge case. I think as long as each of the sessions can have their custom location listed below on the event page and in the ical (the ical was what burned us for Rails Bridge Atlanta) it would actually be alright to display the map with only the main location.
* Alternatively, we could display a marker for each event session location
```ruby
events_helper.rb

def event_locations
  @event.event_sessions.map{ |session| session.custom_location }
end

events/show.html.erb

<% if @event.event_locations.present? && !Rails.env.test? %>
  <%= render "shared/map", locations: @event.event_locations %>
<% end %>

shared/_map.html.erb

<%
  markers = Gmaps4rails.build_markers([locations]) do |location, marker|
    marker.lat location.latitude
    marker.lng location.longitude
  end
%>
```

What do you think?

**Comment by tjgrathwell** [comment]
I hadn't considered using multiple markers, that seems like it could work.

Adding a `custom_location_id` to `event_sessions` seems good. I think all the view logic can probably be driven using the presence or absence of data in that field, kinda like how `childcare_info` works on the Rsvp form.

Is this something you have time to work on in the near-term? There's probably not any big time pressure unless someone has an imminent multi-location workshop they're planning, just trying to get an idea of schedule.

**Comment by KyFaSt** [comment]
I can work on this some today and this week. Or, if you have something more pressing I could try that instead

**Comment by tjgrathwell** [comment]
Nothing else is more pressing per se. Do it!!! :racehorse:

**Comment by kmolo** [comment]
Two comments:
 - I like the idea of multiple pins on single map.  However, one alternate possible resolution to the multiple maps question is to have a map link underneath each session address that will open up to individual maps when clicked

![Alternate Event Page picture with each session having its address underneath the session instead of a single address above all sessions](https://cloud.githubusercontent.com/assets/4886396/9127151/75e87fde-3c6b-11e5-87f4-69936e46058e.png)

...possibly if all sessions are at the same location, keep the original address-above-all-sessions formatting:

![Original Event Page picture with single address above all sessions](https://cloud.githubusercontent.com/assets/1041691/7336042/054847f0-ebad-11e4-89fb-ef068a55d647.png)

---
- if going with plan to have only "main" location displayed, would be nice to have checkbox, so an organizer could pick which address should be the "main" rather than assuming based on which session comes first or last or which location is used most frequently.

**Comment by tjgrathwell** [comment]
This is now implemented as described above, in the commits 59a0a311259fab71596a9e10b15cc699a136502a and dc01d980f6b41966764522407e990a96db8e800b

The feature touches a lot of different areas, so there will probably be some interesting bugs!
